src/client-side-metrics/gcp-metrics-handler.ts

```diff
@@ -0,0 +1,313 @@
+          }
+        ),
+        clientBlockingLatencies: meter.createHistogram(
+          'bigtable.googleapis.com/internal/client/client_blocking_latencies',
+          {
+            description:
+              'Latencies introduced when the client blocks the sending of more requests to the server because of too many pending requests in a bulk operation.',
+            unit: 'ms',
+            advice: {
+              explicitBucketBoundaries: latencyBuckets,
+            },
+          }
+        ),
+      };
+    }
+  }
+
+  /**
+   * Records metrics for a completed Bigtable operation.
+   * This method records the operation latency and retry count, associating them with provided attributes.
+   * @param {OnOperationCompleteData} data Data related to the completed operation.
+   */
+  onOperationComplete(data: OnOperationCompleteData) {
+    this.initialize({
+      projectId: data.projectId,
+      instanceId: data.metricsCollectorData.instanceId,
+      table: data.metricsCollectorData.table,
+      cluster: data.metricsCollectorData.cluster,
+      zone: data.metricsCollectorData.zone,
+    });
+    this.otelMetrics?.operationLatencies.record(data.operationLatency, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      streamingOperation: data.streamingOperation,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.retryCount.add(data.retryCount, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.firstResponseLatencies.record(data.firstResponseLatency, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      clientName: data.clientName,
+    });
+  }
+
+  /**
+   * Records metrics for a completed attempt of a Bigtable operation.
+   * This method records attempt latency, connectivity error count, server latency,
+   * along with the provided attributes.
+   * @param {OnAttemptCompleteData} data Data related to the completed attempt.
+   */
+  onAttemptComplete(data: OnAttemptCompleteData) {
+    this.initialize({
+      projectId: data.projectId,
+      instanceId: data.metricsCollectorData.instanceId,
+      table: data.metricsCollectorData.table,
+      cluster: data.metricsCollectorData.cluster,
+      zone: data.metricsCollectorData.zone,
+    });
+    this.otelMetrics?.attemptLatencies.record(data.attemptLatency, {
```

Comment:
is there a way to get rid of these optionals? It would be good if we could have some more guarantees here.

Maybe turn `initialize` into `getMetrics`, and have it return the metrics object, creating it if needed?

Reply:
Also, nit: I think Metrics is the wrong terminology for these. I think these are [Instruments](https://opentelemetry.io/docs/specs/otel/metrics/#programming-model)

Reply:
> Maybe turn initialize into getMetrics, and have it return the metrics object, creating it if needed?

Yes. I like this. Done.